framework/decode/vulkan_virtual_swapchain.cpp

```diff
@@ -132,81 +135,153 @@ VkResult VulkanVirtualSwapchain::GetSwapchainImagesKHR(PFN_vkGetSwapchainImagesK
+            {
+                blit_queue_family_index = queue_family_index;
+                device_table_->GetDeviceQueue(device, blit_queue_family_index, 0, &initial_blit_queue);
+                if (initial_blit_queue != VK_NULL_HANDLE)
+                {
+                    break;
+                }
+            }
+        }
+
+        // Make sure we have enough storage for each of our tracked components (Command pools,
+        // Command Buffers, Semaphores, etc) as many queue families that are available.
+        // This is because at any point, the application may get a Device queue from that family and
+        // use it during the present.
+        uint32_t start_size = static_cast<uint32_t>(swapchain_info->blit_command_pools.size());
+        uint32_t new_count  = property_count;
+        if (start_size < new_count)
         {
-            device_table_->GetDeviceQueue(
-                device, swapchain_info->queue_family_indices[0], 0, &swapchain_info->blit_queue);
-
-            VkCommandPoolCreateInfo command_pool_create_info = { VK_STRUCTURE_TYPE_COMMAND_POOL_CREATE_INFO };
-            command_pool_create_info.flags =
-                VK_COMMAND_POOL_CREATE_TRANSIENT_BIT | VK_COMMAND_POOL_CREATE_RESET_COMMAND_BUFFER_BIT;
-            command_pool_create_info.queueFamilyIndex = swapchain_info->queue_family_indices[0];
-            VkResult result                           = device_table_->CreateCommandPool(
-                device, &command_pool_create_info, nullptr, &swapchain_info->blit_command_pool);
-            if (result != VK_SUCCESS)
+            swapchain_info->blit_command_pools.resize(new_count);
+            swapchain_info->blit_command_buffers.resize(new_count);
+            swapchain_info->blit_semaphores.resize(new_count);
+            for (uint32_t ii = start_size; ii < new_count; ++ii)
             {
-                return result;
+                swapchain_info->blit_command_pools[ii] = VK_NULL_HANDLE;
             }
 
-            for (uint32_t i = 0; i < capture_image_count; ++i)
+            for (uint32_t queue_family_index = 0; queue_family_index < property_count; ++queue_family_index)
             {
-                VkCommandBufferAllocateInfo allocate_info = { VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO };
-                allocate_info.pNext                       = nullptr;
-                allocate_info.commandPool                 = swapchain_info->blit_command_pool;
-                allocate_info.level                       = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
-                allocate_info.commandBufferCount          = 1;
-                VkCommandBuffer command_buffer            = VK_NULL_HANDLE;
-                result = device_table_->AllocateCommandBuffers(device, &allocate_info, &command_buffer);
-                if (result != VK_SUCCESS)
+                // Create one command pool per queue.
+                if (swapchain_info->blit_command_pools[queue_family_index] == VK_NULL_HANDLE)
```

Comment:
Maybe we could check if the family index of Queue has VK_QUEUE_GRAPHICS_BIT, and then to Create CommandPool, CommandBuffers and Semaphores

Reply:
Its not just used for Graphics, the queues used are also Transfer and compute.  At least for Wolfenstein.  I could initially just do the graphics ones, but I can create they dynamically on a QueuePresent of a new Queue.  Is that what you mean?

Reply:
No. I don't mean creating dynamically. I mean Queue support Graphics, not mean only support Graphics. For Example `Graphics` , `Graphics | Compute | Transfer` ,  `Graphics | Transfer` of Queues that we need. But `Compute | Transfer` , `Transfer`  of Queues that we don't need.

Reply:
I could, but that doesn't mean another queue type doesn't come along and get used.  That's my concern.  But I can change it to do that for now.  @bradgrantham , what do you think?

Reply:
It's ok to keep those queues if they might be used somewhere.

Reply:
@bradgrantham-lunarg , the problem is that `VulkanVirtualSwapchain` only has 1 instance created per run and there could be multiple swapchains per multiple devices.  That's why all the data is stored in the `SwapchainInfo` which is created per swapchain per device.  I think the ideal would be if there was a `Swapchain` class and the `VirtualSwapchain` had been a derived child class and then data could have been stored in the class, but we don't have that design currently. 

Also, as I mentioned a transfer and a compute queue (and even a video queue) could have a `vkQueuePresent` called on it and we would run into the same situation.  For example, a transfer or compute workload might modify the swapchain image with data (let's say an ambient occlusion path). 

Of course, for video-only we probably would currently break. In that case, we'd need to submit an empty command buffer in the video-only queue that triggered a semaphore (let's say SemA), then the BLIT would be added to a graphics queue waiting on SemA to start and signalling SemB when done, then we would put wait for SemB to start the `Present` in the video only queue.

Reply:
**NOTE** I will put a TODO comment about the video-only problem in the code, just so people are aware.

Reply:
Ok, I now use `vkGetPhysicalDeviceSurfaceSupportKHR` to query which Queue Families can perform `VkQueuePresentKHR`.  If they do, then only for those do I create the command pool and buffers.

Reply:
I think we should rename `blit_*` because we don't use `CmdBlitImage` any more, instead of `CmdCopyImage`.